platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp

```diff
@@ -0,0 +1,379 @@
+						fails++;
+					}
+
+					if (count >= 5000) {
+						syslog(LOG_INFO, "==== test 1 : %u failures ====\n", fails);
+						perf_print_counter(_pc_txns);
+						perf_print_counter(_pc_retries);
+						perf_print_counter(_pc_timeouts);
+						perf_print_counter(_pc_crcerrs);
+						perf_print_counter(_pc_protoerrs);
+						perf_print_counter(_pc_uerrs);
+						count = 0;
+					}
+				}
+			}
+			break;
+
+		case 2:
+			syslog(LOG_INFO, "test 2\n");
+			break;
+
+		default:
+			break;
+		}
+
+		break;
+
+	default:
+		ret = -1;
+		break;
+	}
+
+	return ret;
+}
+
+int
+ArchPX4IOSerial::_bus_exchange(IOPacket *_packet)
+{
+
+	int ret = OK;
+
+	uint8_t *packet = (uint8_t *)_packet;
+	size_t send_size = PKT_SIZE(*_packet);
+
+	/* Protect from concurrent access */
+
+	_current_packet = _packet;
+	_packet_cnt = 0;
+
+	/* Incoming packet size too large ? */
+
+	if (send_size > sizeof(*_current_packet)) {
+		perf_count(_pc_protoerrs);
+		return -EINVAL;
+	}
+
+	/* Measure exchange time */
+
+	perf_begin(_pc_txns);
+
+	/* Clear fifos */
+
+	putreg32((UART_FCR_RFIFOR | UART_FCR_XFIFOR), PX4IO_SERIAL_BASE + MPFS_UART_FCR_OFFSET);
+
+	/* Set first interrupt to occur when at least one byte is available */
+
+	putreg32(0, PX4IO_SERIAL_BASE + MPFS_UART_RFT_OFFSET);
+
+	/* Enable received data available interrupt */
+
+	putreg32(UART_IER_ERBFI, PX4IO_SERIAL_BASE + MPFS_UART_IER_OFFSET);
+
+	/* Write package to tx fifo */
+
+	/* Don't allow interrupts during the UART fifo fill; any delay here could timeout the px4io */
+
+	irqstate_t flags = px4_enter_critical_section();
+
+	for (size_t i = 0; i < send_size; i++) {
+		putreg32(packet[i], PX4IO_SERIAL_BASE + MPFS_UART_THR_OFFSET);
+	}
+
+	px4_leave_critical_section(flags);
+
+	/* Wait for response, max 10 ms */
+
+	ret = nxsem_tickwait_uninterruptible(&_completion_semaphore, MSEC2TICK(10));
+
+	if (ret == -ETIMEDOUT) {
+		perf_count(_pc_timeouts);
+		perf_cancel(_pc_txns);		/* don't count this as a transaction */
+	}
+
+	if (ret == OK) {
+		/* Check packet CRC - corrupt packet errors mean IO receive CRC error */
+		uint8_t crc = _current_packet->crc;
+		_current_packet->crc = 0;
+
+		if ((crc != crc_packet(_current_packet)) || (PKT_CODE(*_current_packet) == PKT_CODE_CORRUPT)) {
+			perf_count(_pc_crcerrs);
+			ret = -EIO;
+		}
+	}
+
+	/* Update counters */
+
+	perf_end(_pc_txns);
+
+	return ret;
+}
+int
+ArchPX4IOSerial::_interrupt(int irq, void *context, void *arg)
+{
+	if (arg != nullptr) {
+		ArchPX4IOSerial *instance = reinterpret_cast<ArchPX4IOSerial *>(arg);
+
+		instance->_do_interrupt();
+	}
+
+	return 0;
+}
+
+void
+ArchPX4IOSerial::_do_interrupt()
+{
+
+	/* Get the current status */
+
+	uint32_t status = getreg32(PX4IO_SERIAL_BASE + MPFS_UART_IIR_OFFSET);
+	uint8_t *packet = (uint8_t *)_current_packet;
+	bool first_rcv = _packet_cnt == 0 ? true : false;
+	int ret = OK;
+
+	switch (status & UART_IIR_IID_MASK) {
+
+	case UART_IIR_IID_RECV:
```

Comment:
Where is the interrupt cleared or is this level active, i.e. when FIFO is below threshold IRQ line is deasserted ?